feat(rpc): created EthApiBuilder type

Dhruv-2003 opened the pull request:
closes #14000 
closes #13899

Created a new `EthApiBuilder` type for building a new `EthApi` instance with required arguments & others with a default value

### Note
Currently I have added the default values from what I found in the `build_test_eth_api` & it seemed right to me, let me know if it's not correct and is to be replaced with other value

mattsse commented on crates/rpc/rpc/src/eth/core.rs:
```diff
@@ -450,9 +453,131 @@ where
     }
 }
 
+/// A helper to build the `EthApi` instance.
+#[derive(Debug)]
+pub struct EthApiBuilder<Provider, Pool, Network, EvmConfig>
+where
+    Provider: BlockReaderIdExt,
+{
```
could we move this to builder.rs ? because we likely need a a lot of functions on this one

mattsse commented on crates/rpc/rpc/src/eth/core.rs:
```diff
@@ -450,9 +453,131 @@ where
+    gas_cap: GasCap,
+    max_simulate_blocks: u64,
+    eth_proof_window: u64,
+    blocking_task_pool: BlockingTaskPool,
```
this I'd like to keep as Option just in case the user wants to provide a different pool,

we can do unwrap_or on build

mattsse commented on crates/rpc/rpc/src/eth/core.rs:
```diff
@@ -450,9 +453,131 @@ where
+
+    /// Builds the `EthApi` instance.
+    pub fn build(self) -> EthApi<Provider, Pool, Network, EvmConfig> {
+        EthApi::new(
+            self.provider,
+            self.pool,
+            self.network,
+            self.eth_cache,
+            self.gas_oracle,
+            self.gas_cap,
+            self.max_simulate_blocks,
+            self.eth_proof_window,
+            self.blocking_task_pool,
+            self.fee_history_cache,
+            self.evm_config,
+            self.proof_permits,
+        )
```
could we perhaps flip this, so that ethapi::new uses the ethapibuilder and this does the entire setup?

we likely want to have a separate fn for build ethapiInner type

Dhruv-2003 replied:
just to be sure, u mean i move the actual build logic and initialising `EthApi` to the `EthApiBuilder` function & that allows user to either use `EthApi::new` or `EthApiBuilder` for creating a new instance ?

I would need to extend the provider trait bounds on `EthApi` to ` Provider: BlockReaderIdExt + StateProviderFactory + Clone + Unpin + 'static`

mattsse reviewed (changes requested):
nice, ty

some suggestions/ideas

mattsse reviewed (approved):
cool, ty!

I moved a few things around, and changed my mind about changing the ethapi::new fn for,
need to think about this a bit more first.

Dhruv-2003 commented:
> cool, ty!
> 
> I moved a few things around, and changed my mind about changing the ethapi::new fn for, need to think about this a bit more first.

alright, let me know if I need to make any changes
